### DISCUSSION TITLE
AntDesign.Core Infrastucture Design

### DISCUSSION THREAD

**Discussion by theilgaz** [ant-design-blazor/ant-design-blazor#2085] [Ideas]
Separating the all core functionalities from AntDesign.Blazor project to `AntDesign.Core` is the main idea for this issue. Here are some different approach for roadmap.

@anddrzejb @ElderJames @Zonciu @anranruye @Hona

### Path 1
- Core as a project 
- Component as a project

### Path 2
- All components in a same project

_Originally posted by @theilgaz in https://github.com/ant-design-blazor/ant-design-blazor/issues/2016#issuecomment-961658420_

**Comment by anddrzejb** [comment]
Regardless of the path chosen, the core needs to be extracted first. 

But I will ask again - what is best approach if we would like to offer nuget packages like this:
- AntDesign (all-in-one)
- AntDesign.Core
- AntDesign.Button
- AntDesign.Select
- AntDesign.Tree
- (etc.)

**Reply by theilgaz** [reply]
It seems like more clear and minified. Those packages are not huge. So we can separate them. If someone needs all-in-one package, then we suggest to use `AntDesign` package.

**Reply by anranruye** [reply]
I don't have an experience to publish a nuget package either. I guess we should have a separate project for every single component. And the AntDesign(or AntDesign.All as the name) project/package is just empty and relies on other projects.

**Reply by anddrzejb** [reply]
It looks like everyone is for having every component in a separate project. At least one last person that needs to state his mind is @ElderJames . What do you say?

**Reply by theilgaz** [reply]
> I don't have an experience to publish a nuget package either. I guess we should have a separate project for every single component. And the AntDesign(or AntDesign.All as the name) project/package is just empty and relies on other projects.

Don't worry about NuGet Packages. It's very simple step. 

@ElderJames please confirm one more time, Shunjie. If all of us agree, then we can start building `AntDesign.Core` and components together.

**Reply by ElderJames** [reply]
I'm in favor of separating out `AntDesign.Core` in order to use it for community-contributed components, but I'd like to leave the antdesign component alone for now.

**Reply by knight1219** [reply]
Will the CSS be separated as well? So each independent control package will have its own CSS and JS files to add?

**Reply by anddrzejb** [reply]
Possibly. Regarding js - I am not yet sure. To be honest I wanted to play around with js isolation, but didn't get a chance yet. With js isolation we might not even need to separate the js into different modules. The problem is that AntBlazor does also support net3.1. No isolation there would mean more overhead for users (adding script tags for each component).

**Reply by zxyao145** [reply]
I think splitting components is a good practice. In addition to effectively reducing the size of the package, I think there are the following points.

**First**, we can use different technology stacks in different libraries, such as CSS + JS, SCSS + TS, LESS+ ts, etc. Although LESS is used in AntDesign, I'm sorry, I don't know how to use LESS, I usually choose to use SCSS.

**Secondly**, we can (but do not have to) develop a template for community library, such as integrating Gulp or Webpack, or even code style, so that as long as contributors adopt this template, they can easily contribute to enrich the community.

**However**, there may be conflicts between different community components, which is an unavoidable disadvantage.

**Reply by theilgaz** [reply]
I prefer SASS too. Ant Design needs SASS, instead of LESS. It's a common issue as I searched. There are many different tech stacks user communities are in pain because of LESS. The recent I searched is Next.Js+AntD combination. Using custom styles for primary and other component styles can be painful. I think SASS is more suitable for our environment. Also I researched other Blazor components which are using SASS.
